tests/components/broadlink/test_device.py

```diff
@@ -0,0 +1,389 @@
+    ) as mock_init:
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_SETUP_RETRY
+    assert mock_api.auth.call_count == 1
+    assert mock_api.check_sensors.call_count == 1
+    assert mock_forward.call_count == 0
+    assert mock_init.call_count == 0
+
+
+async def test_device_setup_get_fwversion_broadlink_exception(hass):
+    """Test we load the device even if we cannot read the firmware version."""
+    device = get_device("Office")
+    mock_api = device.get_mock_api()
+    mock_api.get_fwversion.side_effect = blke.BroadlinkException()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api), patch.object(
+        hass.config_entries, "async_forward_entry_setup"
+    ) as mock_forward:
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_LOADED
+    forward_entries = {c[1][1] for c in mock_forward.mock_calls}
+    domains = get_domains(mock_api.type)
+    assert mock_forward.call_count == len(domains)
+    assert forward_entries == domains
+
+
+async def test_device_setup_get_fwversion_os_error(hass):
+    """Test we load the device even if we cannot read the firmware version."""
+    device = get_device("Office")
+    mock_api = device.get_mock_api()
+    mock_api.get_fwversion.side_effect = OSError()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api), patch.object(
+        hass.config_entries, "async_forward_entry_setup"
+    ) as mock_forward:
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_LOADED
+    forward_entries = {c[1][1] for c in mock_forward.mock_calls}
+    domains = get_domains(mock_api.type)
+    assert mock_forward.call_count == len(domains)
+    assert forward_entries == domains
+
+
+async def test_device_setup_registry(hass):
+    """Test we register the device and the entries correctly."""
+    device = get_device("Office")
+    mock_api = device.get_mock_api()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    device_registry = mock_device_registry(hass)
+    entity_registry = mock_registry(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api):
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+        await hass.async_block_till_done()
+
+    assert len(device_registry.devices) == 1
+
+    device_entry = device_registry.async_get_device(
+        {(DOMAIN, mock_entry.unique_id)}, set()
+    )
+    assert device_entry.identifiers == {(DOMAIN, device.mac)}
+    assert device_entry.name == device.name
+    assert device_entry.model == device.model
+    assert device_entry.manufacturer == device.manufacturer
+    assert device_entry.sw_version == device.fwversion
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6042836)